aten/src/ATen/native/cuda/ForeachTensorAddScalar.cu

```diff
@@ -0,0 +1,100 @@
+// of a __device__ lambda not have internal linkage.
+
+namespace at { namespace native {
+
+namespace {
+
+template<typename x_t, typename out_t>
+struct AddScalarFunctor {
+    __device__ void operator() (
+        int chunk_size,
+        TensorListMetadata<2>& tl,
+        x_t scalar) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
+
+            x_t* x = (x_t*)tl.addresses[0][tensor_loc];
+            x += chunk_idx * chunk_size;
+
+            out_t* out = (out_t*)tl.addresses[1][tensor_loc];
+            out += chunk_idx * chunk_size;
+
+            n -= chunk_idx * chunk_size;
+
+            x_t r_x[kILP];
+            out_t r_out[kILP];
+
+            // to make things simple, we put aligned case in a different code path
+            if(n % kILP == 0 && chunk_size % kILP == 0 && is_aligned(x) && is_aligned(out)) {
+                for(int i_start = threadIdx.x; i_start * kILP < n && i_start * kILP < chunk_size; i_start += blockDim.x) {
+                    // load
+                    load_store(r_x, x, 0 , i_start);
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+                    // store
+                    load_store(out, r_out, i_start, 0);
+                }
+            }
+            else {
+                // Non-divergent exit condition for __syncthreads, not necessary here
+                for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = 0;
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size)
+                            out[i] = r_out[ii];
+                    }
+                }
+            }
+        }
+};
+
+} // namespace
+
+std::vector<Tensor> foreach_tensor_add_scalar_kernel_cuda(TensorList tensors, Scalar scalar) {
+    TORCH_CHECK(std::all_of(tensors.begin(), tensors.end(), [] (const Tensor& t) {
+        return t.layout() == at::kStrided;
+    }), "Only tensors with strided layouts are supported.");
+
+    TORCH_CHECK(std::all_of(tensors.begin(), tensors.end(), [] (const Tensor& t) {
+        return t.is_non_overlapping_and_dense();
+    }), "Only non overlapping and dense tensors are supported.");
```

Comment:
This sort of error checking will be common across all kernels.  Therefore, it belongs in the [multi_tensor_apply function itself](https://github.com/pytorch/pytorch/pull/41554/files#diff-ef767cf918d40173403532003ec78ac8R23), not in specific kernel functions.

The checks here are ok for scalar addition to a single tensor list, but not for multi tensor apply ops in which multiple tensor lists interact (e.g. adding a list of tensors to another list of tensors)

For ops where multiple tensor lists interact, what we want to check is:
- All tensors are kStrided.
- All tensors are non overlapping and dense.
- All TensorLists have the same length/number of tensors.
- For each tensor, its partner(s) in the other list(s) must have the same `size()` and `strides()`, ie
```C++
// list_of_tensor_lists[0] is the first list, list_of_tensor_lists[1] is the second list, etc.
// list_of_tensor_lists[0][0] is a Tensor that will interact with list_of_tensor_lists[1][0],
// list_of_tensor_lists[2][0], etc.
for (const TensorList& tl: list_of_tensor_lists {
  for (int i = 0; i < tl.size(); i++) {
    // checks every tensor against its partner in the first list
    tl[i].sizes() = list_of_tensor_lists[0][i].sizes();
    tl[i].strides() = list_of_tensor_lists[0][i].strides();
  }
}
```